File: leetcode/144. Binary Tree Preorder Traversal.cpp
#include <iostream>
#include <unordered_set>
#include <algorithm>
#include <vector>
#include <string>
#include <numeric>
#include <queue>
#include <stack>
using namespace std;



// Definition for a binary tree node.
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

 class Solution {
 public:
	 vector<int> preorderTraversal(TreeNode* root) {
		 if (!root) return{};
		 vector<int> res;
		 stack<TreeNode*> s{ { root } };
		 while (!s.empty()) {
			 TreeNode *t = s.top(); s.pop();
			 res.push_back(t->val);
			 if (t->right) s.push(t->right);
			 if (t->left) s.push(t->left);
		 }
		 return res;
	 }
 };

 class Solution1 {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> s;
        TreeNode *p = root;
        while (!s.empty() || p) {
            if (p) {
                s.push(p);
                res.push_back(p->val);
                p = p->left;
            } else {
                TreeNode *t = s.top(); s.pop();
                p = t->right;
            }
        }
        return res;
    }
};

File: leetcode/144. Binary Tree Preorder Traversal.py
'''Given a binary tree, return the preorder traversal of its nodes' values.

Example:

Input: [1,null,2,3]
   1
    \
     2
    /
   3

Output: [1,2,3]
Follow up: Recursive solution is trivial, could you do it iteratively?
'''

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode):
        ##先续遍历
        res = []
        if root ==None:
            return res
        q = []
        q.append(root)
        while len(q)!=0:
            cur = q.pop(0)
            res.append(cur.val)
            if cur.right:
                q.insert(0,cur.right)
            if cur.left:
                q.insert(0,cur.left)
        return res

        